app-drawer/app-drawer.html

```diff
@@ -260,75 +260,198 @@
+        if (isVelocityPositive) {
+          this._velocity = Math.max(this._velocity, this.MIN_TRANSITION_VELOCITY);
+        } else {
+          this._velocity = Math.min(this._velocity, -this.MIN_TRANSITION_VELOCITY);
+        }
+
+        // Using this timing function, calculate the amount of time needed to finish the
+        // transition.
+        const timingFunction = 'cubic-bezier(.667, 1, .667, 1)';
+        const initialSlope = 1.5;
+        var duration;
+        if (isClosingLeft) {
+          duration = (initialSlope * -(x + drawerWidth) / this._velocity) + 'ms';
+        } else if (isClosingRight) {
+          duration = (initialSlope * (drawerWidth - x) / this._velocity) + 'ms';
+        } else {
+          duration = (initialSlope * -x / this._velocity) + 'ms';
+        }
+        this.$.contentContainer.style.transitionDuration = duration;
+        this.$.contentContainer.style.transitionTimingFunction = timingFunction;
+        this.$.scrim.style.transitionDuration = duration;
+        this.$.scrim.style.transitionTimingFunction = timingFunction;
+
+        // Transform instead of toggling opened to avoid calling Polymer setters.
+        if (isClosingLeft) {
+          this._translateDrawer(-drawerWidth);
+        } else if (isClosingRight) {
+          this._translateDrawer(drawerWidth);
+        } else {
+          this._translateDrawer(0);
+        }
+      },
+
+      _transitionEnd: function() {
+        // Only set the opened state and reset transition timing function if this
+        // transition end event was the end of a fling transition.
+        if (this.$.contentContainer.style.transitionTimingFunction) {
+          if (this._velocity < 0) {
+            this.opened = this.position === 'right';
+          } else {
+            this.opened = this.position === 'left';
+          }
+
+          this.$.contentContainer.style.transitionTimingFunction = '';
+          this.$.scrim.style.transitionTimingFunction = '';
+          this._resetStyleAttributes();
+        }
+
+        this._updateDocScroll();
+      },
+
+      _resetStyleAttributes: function() {
+        this.$.contentContainer.style.transitionDuration = '';
+        this.$.scrim.style.transitionDuration = '';
+        this.$.scrim.style.visibility = '';
+        this.$.scrim.style.opacity = '';
+        this.transform('', this.$.contentContainer);
+      },
+
+      /**
+       * @return {boolean} True if the given position is valid, false otherwise.
+       */
       _translateDrawer: function(x) {
-        this.transform('translate3d(' + x + 'px, 0, 0)', this.$.contentContainer);
+        var drawerWidth = this.getWidth();
+        var clampedX = x;
+
+        if (this.position === 'left') {
+          clampedX = Math.max(-drawerWidth, Math.min(x, 0));
+          this.$.scrim.style.opacity = 1 + clampedX / drawerWidth;
+        } else {
+          clampedX = Math.max(0, Math.min(x, drawerWidth));
```

Comment:
would it be possible to negate `clampedX` here.  e.g. `clampedX= -Math.max(0, Math.min(x, drawerWidth))`. That way, we will only need one `this.$.scrim.style.opacity = 1 + clampedX / drawerWidth;` and we could have a ternary operation.